runtime/src/main/java/org/corfudb/runtime/FastSmrMapsLoader.java

```diff
@@ -0,0 +1,209 @@
+            return runtime.getAddressSpaceView().read(address);
+        } else {
+            return runtime.getAddressSpaceView().fetch(address);
+        }
+    }
+
+    /** Create a new object SMRMap as recipient of SMRUpdates (if doesn't exist yet)
+     *
+     * @param streamId
+     */
+    private void createSmrMapIfNotExist(UUID streamId) {
+        if (!runtime.getObjectsView().getObjectCache().containsKey(streamId)) {
+            runtime.getObjectsView().build()
+                    .setStreamID(streamId)
+                    .setType(SMRMap.class)
+                    .open();
+        }
+    }
+
+    /** Extract log entries from logData and update the Corfu Objects
+     *
+     * @param logData LogData received from Corfu server.
+     */
+    private void processLogData(ILogData logData) {
+        LogEntry logEntry;
+        try {
+            logEntry = logData.getLogEntry(runtime);
+        } catch (Exception e) {
+            return;
+        }
+
+        if (logEntry.getType() == LogEntry.LogEntryType.SMR) {
+            // Just one stream, always
+            UUID streamId = logData.getStreams().iterator().next();
+            applySmrEntryToStream(streamId, (SMREntry) logEntry);
+        }
+        else if (logEntry.getType() == LogEntry.LogEntryType.MULTIOBJSMR) {
+            MultiObjectSMREntry multiObjectLogEntry = (MultiObjectSMREntry) logEntry;
+            multiObjectLogEntry.getEntryMap().forEach((stream, multiSmrEntry) -> {
+                multiSmrEntry.getSMRUpdates(stream).forEach((smrEntry) -> {
+                    applySmrEntryToStream(stream, smrEntry);
+                });
+            });
+
+        }
+    }
+
+    private void findAndSetLogHead() {
+        logHead = 0;
+    }
+
+    private void findAndSetLogTail() {
+        logTail = runtime.getSequencerView().nextToken(Collections.emptySet(), 0).getTokenValue();
+    }
+
+    /** Initialize log head and log tails
+     *
+     * If logHead and logTail has not been initialized by
+     * the user, initialize to default.
+     *
+     */
+    private void initializeHeadAndTails() {
+        if (logHead < 0) {
+            findAndSetLogHead();
+        }
+
+        if (logTail < 0) {
+            findAndSetLogTail();
+        }
+    }
+
+
+
+    /** Entry point to load the SMRMaps in memory.
+     *
+     * When this function returns, the maps are fully loaded.
+     *
+     */
+    public void loadMaps() {
+        initializeHeadAndTails();
+
+        for (long i = logHead; i <= logTail; i++) {
+            ILogData currentLogData = getLogData(i);
+
+            switch (currentLogData.getType()) {
+                case DATA:
+                    processLogData(currentLogData);
+                    break;
+                case HOLE:
+                    break;
+                case TRIMMED:
+                    log.warn("Reading Trimmed Entry at address " + i +
+                            " Head of the log is set to " + logHead);
```

Comment:
There is an option in our log format to print the method as well. Should we have the developers write this down if we can have it for free? https://logback.qos.ch/manual/layouts.html

The only concern is performance. Using reflection to get the name of the function might be costly.